src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala

```diff
@@ -56,31 +59,115 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
+            if (headerIds.nonEmpty) {
+              val maxElements = desiredSizeOfExpectingHeaderQueue - deliveryTracker.requestedSize
+              requestDownload(
+                maxElements,
+                minElementsPerBucket = 50,
+                maxElementsPerBucket = 400
+              )(Option(getPeersForDownloadingHeaders(remote))) { howManyPerType =>
+                Map(Header.modifierTypeId -> headerIds.reverse.filter(downloadRequired(historyReader)).take(howManyPerType))
+              }
+            }
+          case Equal => // does nothing for `Equal`
+        }
+      case _ =>
+    }
+  }
 
-        toDownload.groupBy(_._1).foreach(ids => requestDownload(ids._1, ids._2.map(_._2)))
+  /**
+    * Headers should be downloaded from an Older node, it is triggered by received sync message from an older node
+    * @param callingPeer that can be used to download headers, it must be Older
+    * @return available peers to download headers from together with the state/origin of the peer
+    */
+  private def getPeersForDownloadingHeaders(callingPeer: ConnectedPeer): (PeerSyncState, Iterable[ConnectedPeer]) =
+    statusTracker.peersByStatus.get(Older).map(PeerSyncState.Older -> _).getOrElse(PeerSyncState.OlderCalling -> Iterable(callingPeer))
+
+  /**
+    * Other persistent modifiers besides headers should be downloaded from either Older or Equal node, with fallback to Unknown or Fork
+    * @return available peers to download persistent modifiers from together with the state/origin of the peer
+    */
+  private def getPeersForDownloadingBlocks: Option[(PeerSyncState, Iterable[ConnectedPeer])] = {
+    val peersByStatus = statusTracker.peersByStatus
+    Option(peersByStatus.getOrElse(Older, Iterable.empty) ++ peersByStatus.getOrElse(Equal, Iterable.empty))
+      .filter(_.nonEmpty)
+      .map(PeerSyncState.OlderOrEqual -> _)
+      .orElse {
+        Option(peersByStatus.getOrElse(Unknown, Iterable.empty) ++ peersByStatus.getOrElse(Fork, Iterable.empty))
+          .filter(_.nonEmpty)
+          .map(PeerSyncState.UnknownOrFork -> _)
       }
   }
 
-  // todo: this method is just a copy of the ancestor from Scorex, however, smarter logic is needed, not just
-  //  asking from a random peer
-  override protected def requestDownload(modifierTypeId: ModifierTypeId, modifierIds: Seq[ModifierId]): Unit = {
-    deliveryTracker.setRequested(modifierIds, modifierTypeId, None)
-    val msg = Message(requestModifierSpec, Right(InvData(modifierTypeId, modifierIds)), None)
-    networkControllerRef ! SendToNetwork(msg, SendToRandom)
-  }
+  protected def requestDownload(maxElements: Int, minElementsPerBucket: Int, maxElementsPerBucket: Int)
+                               (getPeers: => Option[(PeerSyncState, Iterable[ConnectedPeer])])
+                               (fetchMax: Int => Map[ModifierTypeId, Seq[ModifierId]]): Unit =
+    getPeers
+      .foreach { case (peerStatus, peers) =>
```

Comment:
Why to ask every status for modifier ids ?

Reply:
`getPeers` returns `Option` of peers of certain `PeerSyncState`